File: judgels-backends/judgels-grader-api/src/main/java/judgels/gabriel/api/Verdict.java
package judgels.gabriel.api;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

@JsonDeserialize(using = Verdict.VerdictDeserializer.class)
public enum Verdict {
    @JsonProperty ACCEPTED("AC", "Accepted"),
    @JsonProperty OK("OK", "OK"),
    @JsonProperty SKIPPED("SKP", "Skipped"),
    @JsonProperty WRONG_ANSWER("WA", "Wrong Answer"),
    @JsonProperty TIME_LIMIT_EXCEEDED("TLE", "Time Limit Exceeded"),
    @JsonProperty RUNTIME_ERROR("RTE", "Runtime Error"),
    @JsonProperty COMPILATION_ERROR("CE", "Compilation Error"),
    @JsonProperty INTERNAL_ERROR("!!!", "Internal Error"),
    @JsonProperty PENDING("?", "Pending");

    private String code;
    private String name;

    Verdict(String code, String name) {
        this.code = code;
        this.name = name;
    }

    public String getCode() {
        return code;
    }

    public String getName() {
        return name;
    }

    @JsonValue
    Map<String, String> toValue() {
        Map<String, String> val = new HashMap<>();
        val.put("code", code);
        return val;
    }

    static class VerdictDeserializer extends JsonDeserializer<Verdict> {
        @Override
        public Verdict deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
            JsonNode node = jp.readValueAsTree();
            String code = node.get("code").asText();
            return Verdicts.fromCode(code);
        }
    }
}

File: judgels-client/src/modules/api/gabriel/verdict.js
import { Intent } from '@blueprintjs/core';

export const VerdictCode = {
  AC: 'AC',
  CE: 'CE',
  WA: 'WA',
  TLE: 'TLE',
  RTE: 'RTE',
  PND: '?',
  ERR: '!!!',
  SKP: 'SKP',
};

const verdictIntentsMap = {
  [VerdictCode.AC]: Intent.SUCCESS,
  [VerdictCode.WA]: Intent.DANGER,
  [VerdictCode.CE]: Intent.NONE,
  [VerdictCode.PND]: Intent.NONE,
  [VerdictCode.ERR]: Intent.NONE,
  [VerdictCode.SKP]: Intent.NONE,
};

const verdictDisplayNamesMap = {
  [VerdictCode.AC]: 'Accepted',
  [VerdictCode.WA]: 'Wrong Answer',
  [VerdictCode.TLE]: 'Time Limit Exceeded',
  [VerdictCode.RTE]: 'Runtime Error',
  [VerdictCode.CE]: 'Compilation Error',
  [VerdictCode.PND]: 'Pending',
  [VerdictCode.ERR]: 'Internal Error',
  [VerdictCode.SKP]: 'Skipped',
};

export function getVerdictIntent(code) {
  return verdictIntentsMap[code] || Intent.WARNING;
}

export function getVerdictDisplayName(code) {
  return verdictDisplayNamesMap[code] || code;
}
